crates/bevy_asset/src/loader.rs

```diff
@@ -541,6 +542,105 @@ impl<'a> LoadContext<'a> {
         Ok(loaded_asset)
     }
+
+    /// Loads the asset at the given `path` directly from the provided `reader`. This is an async function that will wait until the asset is fully loaded before
+    /// returning. Use this if you need the _value_ of another asset in order to load the current asset, and that value comes from your [`Reader`].
+    /// For example, if you are deriving a new asset from the referenced asset, or you are building a collection of assets. This will add the `path` as a
+    /// "load dependency".
+    ///
+    /// If the current loader is used in a [`Process`] "asset preprocessor", such as a [`LoadAndSave`] preprocessor,
+    /// changing a "load dependency" will result in re-processing of the asset.
+    ///
+    /// [`Process`]: crate::processor::Process
+    /// [`LoadAndSave`]: crate::processor::LoadAndSave
+    pub async fn load_direct_with_reader<'b>(
+        &mut self,
+        reader: &mut Reader<'_>,
+        path: impl Into<AssetPath<'b>>,
+    ) -> Result<ErasedLoadedAsset, LoadDirectError> {
+        let path = path.into().into_owned();
+
+        let loader = self
+            .asset_server
+            .get_path_asset_loader(&path)
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: path.clone(),
+                error: error.into(),
+            })?;
+
+        let meta = loader.default_meta();
+
+        let loaded_asset = self
+            .asset_server
+            .load_with_meta_loader_and_reader(
+                &path,
+                meta,
+                &*loader,
+                reader,
+                false,
+                self.populate_hashes,
+            )
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: path.clone(),
+                error,
+            })?;
+
+        let info = loaded_asset
+            .meta
+            .as_ref()
+            .and_then(|m| m.processed_info().as_ref());
+
+        let hash = info.map(|i| i.full_hash).unwrap_or_default();
+
+        self.loader_dependencies.insert(path, hash);
+
+        Ok(loaded_asset)
+    }
+
+    /// Retrieves a handle for the asset at the given path and adds that path as a dependency of the asset.
+    /// If the current context is a normal [`AssetServer::load`], an actual asset load will be kicked off immediately, which ensures the load happens
+    /// as soon as possible.
+    /// "Normal loads" kicked from within a normal Bevy App will generally configure the context to kick off loads immediately.  
+    /// If the current context is configured to not load dependencies automatically (ex: [`AssetProcessor`](crate::processor::AssetProcessor)),
+    /// a load will not be kicked off automatically. It is then the calling context's responsibility to begin a load if necessary.
+    pub fn load_with_reader<'b, A: Asset>(
+        &mut self,
+        reader: Box<Reader<'static>>,
+        path: impl Into<AssetPath<'b>>,
```

Comment:
I would have expected `load_with_reader` to not have a path. Hmm sorry for the back and forth. Really sorry :sweat:, I didn't realize there would be additional complex design decisions involved. I think we are fine with only the `load_direct_with_reader`. We can add the rest in a future PR.

Reply:
Thats a very good PR you linked.

`load_with(_settings_and)_reader` really shouldnt take a path. The path is only used as a type hint to know what loader to use and if #10153 gets merged should not be necessary anymore. The current implementation also registers the path on the `AssetServer` and returns a strong `Handle` with a path. This doesnt feel right since the asset is not actually backed by any file at that path. An asset loaded with a reader should not have a path associated with it.

It also creats a new footgun when reloading that asset. Since it isnt backed by an actual file the reload wouldnt do anything. Or worse it could reload a different file that is actually at that path.

Reply:
I've reverted the change which added those extra methods, since I think #10153 would substantially change their API (if merged). I agree that they could be added with an additional PR at a future date if still desired.